matches/matches.go

```diff
@@ -2,19 +2,161 @@ package main
 
 import (
 	"encoding/json"
+	"fmt"
 	"log"
 	"net/http"
+	"strconv"
 
+	"github.com/TempleEight/spec-golang/matches/dao"
+	valid "github.com/asaskevich/govalidator"
 	"github.com/gorilla/mux"
 )
 
 func main() {
 	r := mux.NewRouter()
-	r.HandleFunc("/matches", matchGetHandler).Methods(http.MethodGet)
+	r.HandleFunc("/matches/{id}", matchListHandler).Methods(http.MethodGet)
+	r.HandleFunc("/match/{id}", matchGetHandler).Methods(http.MethodGet)
+	r.HandleFunc("/match", matchCreateHandler).Methods(http.MethodPost)
+	r.HandleFunc("/match/{id}", matchDeleteHandler).Methods(http.MethodDelete)
+	r.HandleFunc("/match", matchUpdateHandler).Methods(http.MethodPatch)
```

Comment:
This is a `PUT`, not a `PATCH`: [see here](https://rapidapi.com/blog/put-vs-patch/)

Reply:
That link supports that the current implementation should be a `PATCH`, a partial update. as the `matchedOn` field is not changed. Also it doesn't create a new entry if it doesn't exist.

However leaving the `matchedOn` field unchanged (and not giving the option to update it) seems wrong.

I agree that this PR is far too large, there are lots of small decisions being made here and things being overlooked.

Reply:
I think this one is quite ambiguous since a request could never update the `matchedOn` field, which means the update request body matches that of the `POST` request. To an end user of this service this the would suggest entire resource is updated, hence I suggested `PUT`. 

`PATCH` would also imply that the user could only provide certain fields to the body and only these fields would be updated: this isn't the case at all, since we update all fields using the request contents.

With regards to creating the entity if it doesn't exist, the RFC has some interesting wording around that implying it's a choice rather than a requirement:

> If the Request-URI does not point to an existing resource, and that URI is capable of being defined as a new resource by the requesting user agent, the origin server **can** create the resource with that URI.

We can leave it as is now, but I think I'd actually support removing the update functionality all together.